### PR TITLE
Added results retrieval approach

### PR DESCRIPTION
**Pull request by dotsdl**
Adds use of `ObjectStore` alongside the `Neo4jStore` (state store) by the user-facing `AlchemiscaleClient`/`AlchemiscaleAPI` and compute-facing `AlchemiscaleComputeClient`/`AlchemiscaleComputeAPI` for `ProtocolDAGResult` storage and retrieval.

Integration tests for the two clients and storage components are provided. We don't have focused tests on the APIs themselves yet; we need a less painful way of building tests out for these than we currently have.

### CODE REVIEW

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few quick comments and questions from just scanning the code. I need to sit down with a diagram and do a few drawings before I can think clearly about the design but looks great so far. :)

File: fah_alchemy/models.py
Comment (hmacdope) [inline fah_alchemy/models.py:65]:
```suggestion
    def specific(self) -> bool:
```

File: fah_alchemy/storage/objectstore.py
Comment (hmacdope) [inline fah_alchemy/storage/objectstore.py:16]:
```suggestion
def get_s3os(settings: S3ObjectStoreSettings, endpoint_url=None) -> S3ObjectStore:
```

File: fah_alchemy/storage/objectstore.py
Comment (hmacdope) [inline fah_alchemy/storage/objectstore.py:184]:
```suggestion
    ) -> ObjectStoreRef:
```

File: fah_alchemy/storage/statestore.py
Comment (hmacdope) [inline fah_alchemy/storage/statestore.py:854]: Am I correct in thinking we only delete the relationship but leave the node intact incase there is another task depending on it? Just checking my understanding.

**Review by hmacdope — CHANGES_REQUESTED** [review]
Few more comments, mostly just tiny nitpicks on typing etc. 

Design looks clean and elegant. :)

File: fah_alchemy/interface/api.py
Comment (dotsdl) [reply fah_alchemy/interface/api.py:146]: `sk` means `ScopedKey` here; it's a convention we've been using in the test suite to indicate that we're dealing with the `ScopedKey` of the object, not the object itself.

Does that work for you?

Comment (hmacdope) [reply fah_alchemy/interface/api.py:147]: Yes sorry wrote this before seeing  sk everywhere. Perfectly fine :)

File: fah_alchemy/storage/objectstore.py
Comment (dotsdl) [reply fah_alchemy/storage/objectstore.py:205]: Thanks! Got it!

File: fah_alchemy/storage/objectstore.py
Comment (dotsdl) [reply fah_alchemy/storage/objectstore.py:134]: Ah thank you! This was for debug!

File: fah_alchemy/interface/api.py
Comment (dotsdl) [reply fah_alchemy/interface/api.py:147]: No problem, thanks for asking!

File: fah_alchemy/storage/statestore.py
Comment (dotsdl) [reply fah_alchemy/storage/statestore.py:886]: In general we won't ever delete tasks, exactly for that reason, and also because we attach our `ProtocolDAGResultRef`s to them on completion or failure.

**Comment by dotsdl** [comment]
@hmacdope after I get this PR settled, I'll articulate our outstanding needs for writing up a clear task status lifecycle. It will be much clearer where we need attention by then, and I'd like you to lead that effort. Does that sound interesting to you?

**Comment by hmacdope** [comment]
> @hmacdope as a heads up, after I get this PR settled, I'll articulate our outstanding needs for writing up a clear task status lifecycle. It will be much clearer where we need attention by then, and I'd like you to lead that effort. Does that sound interesting to you?

Yep for sure. I'll hold off on documenting in detail in #80 in that case?

File: fah_alchemy/compute/service.py
Comment (dotsdl) [reply fah_alchemy/compute/service.py:136]: Done!

File: fah_alchemy/interface/api.py
Comment (dotsdl) [reply fah_alchemy/interface/api.py:207]: Not this one, actually; this is technically the string form here.

File: fah_alchemy/interface/api.py
Comment (dotsdl) [reply fah_alchemy/interface/api.py]: Done!

File: fah_alchemy/interface/api.py
Comment (dotsdl) [reply fah_alchemy/interface/api.py:251]: Not this one; this is a string here.

File: fah_alchemy/interface/client.py
Comment (dotsdl) [reply fah_alchemy/interface/client.py:187]: Done!

File: fah_alchemy/compute/service.py
Comment (dotsdl) [reply fah_alchemy/compute/service.py:165]: Ah, it does now! Thanks!

File: fah_alchemy/interface/client.py
Comment (dotsdl) [reply fah_alchemy/interface/client.py:50]: It does now!

File: fah_alchemy/interface/client.py
Comment (dotsdl) [reply fah_alchemy/interface/client.py:33]: We raise an exception for cases that aren't specific now.

**Comment by dotsdl** [comment]
@hmacdope this is ready for a final review from you. We've seen some changes upstream in `openfe` for our `openfe-benchmarks` dependency, so those may break things here yet.

**Comment by dotsdl** [comment]
CI failing; will have to diagnose in the morning.

**Comment by dotsdl** [comment]
@hmacdope finally have tests passing and all comments resolved!

File: alchemiscale/models.py
Comment (hmacdope) [inline alchemiscale/models.py:39]: This is "glob-matching" as separate from scope-glob matching right? 

This dissallows "my*org-blah-blah" ?

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [inline alchemiscale/storage/statestore.py:827]: Are there any desired behaviour changes here I should be aware of for #80 ?

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (hmacdope) [inline alchemiscale/tests/integration/storage/test_statestore.py:478]: I think ordering in this case will become non-deterministic when #78 is in? No change needs to be made but I may need to modify the test.

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (hmacdope) [inline alchemiscale/tests/integration/storage/test_statestore.py:495]: Again query on order for #78 as this will likely break. No change needs to be made but I may need to modify the test.

**Review by hmacdope — APPROVED** [review]
Looks great to me @dotsdl. Thanks for the extensive testing. Just a few queries on behaviour for my own understanding but  I think you should go ahead and merge.  :)

File: alchemiscale/storage/statestore.py
Comment (hmacdope) [reply alchemiscale/storage/statestore.py:827]: Ah I see you have just changed from != complete to != complete or NULL. Sorry got a bit confused by the diff.

File: alchemiscale/models.py
Comment (dotsdl) [reply alchemiscale/models.py:39]: Correct! We don't want to support e.g. `"my*org-blah-blah"`, so we scream here if that is attempted.

File: alchemiscale/storage/statestore.py
Comment (dotsdl) [reply alchemiscale/storage/statestore.py:827]: Oh yeah, the diff is ugly here. I think for #80 the picture is simplified greatly, so I think we're in good shape there. Will keep that in mind when I review #80 after merging this.

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (dotsdl) [reply alchemiscale/tests/integration/storage/test_statestore.py:478]: Indeed, we'll ditch the ordering tests here for #78/#80.

File: alchemiscale/tests/integration/storage/test_statestore.py
Comment (dotsdl) [reply alchemiscale/tests/integration/storage/test_statestore.py:495]: Agreed, we'll be able to delete ordering tests in #80 once this is merged.

**Comment by dotsdl** [comment]
Thanks for the review @hmacdope! Merging!

**Pull request merged**
